src/main/cpp/subsystems/Arm.cpp

```diff
@@ -1,41 +1,123 @@
 #include "subsystems/Arm.h"
+#include "Parameters.h"
+#include "ui/UserInterface.h"
 
+//pulls PID values from the parameters.toml file
 bool Arm::Init() {
-    arm_pid_ = std::make_shared<frc::PIDController>(0, 0, 0);
+    
+    double arm_kP = RobotParams::GetParam("arm.lift_pid.kP", 0.005);
+    double arm_kI = RobotParams::GetParam("arm.lift_pid.kI", 0.0);
+    double arm_kD = RobotParams::GetParam("arm.lift_pid.kD", 0.0);
 
-    inches_pid_ = std::make_shared<frc::PIDController>(0, 0, 0);
+    double extend_kP = RobotParams::GetParam("arm.extend_pid.kP", 0.005);
+    double extend_kI = RobotParams::GetParam("arm.extend_pid.kI", 0.0);
+    double extend_kD = RobotParams::GetParam("arm.extend_pid.kD", 0.0);
+
+    arm_pid_ = std::make_shared<frc::PIDController>(arm_kP, arm_kI, arm_kD);
+
+    inches_pid_ = std::make_shared<frc::PIDController>(extend_kP, extend_kI, extend_kD);
 
     arm_lift_output_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/lift_output");
     arm_lift_enc_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/lift_enc");
+    arm_lift_setpoint_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/lift_setpoint");
+
+    arm_extend_output_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/extend_output");
+    arm_extend_enc_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/extend_enc");
+    arm_extend_setpoint_log_ = wpi::log::DoubleLogEntry(TeamOKC::log, "/arm/extend_setpoint");
+
+    this->arm_pid_->SetSetpoint(0);
+    this->inches_pid_->SetSetpoint(0);
+
+    lift_limit = RobotParams::GetParam("arm.lift_limit", 100);
+    extend_limit = RobotParams::GetParam("arm.extend_limit", 100);
+     
 
     return true;
 }
-bool Arm::SetControlMode(const ArmMode &mode){
+
+bool Arm::SetControlMode(const ControlMode &mode){
     mode_= mode;
 
     return true;
 }
+
 bool Arm::SetDegrees(double degrees) {
-    this->arm_pid_->SetSetpoint(degrees);
+    
+    OKC_CHECK(this->arm_pid_ != nullptr);
+
+    // limit the lift PID setpoint to the actual hardware limits
+    if (degrees < lift_limit) {
+        if (degrees > -lift_limit) {
+            this->arm_pid_->SetSetpoint(degrees);
+        } else {
+            this->arm_pid_->SetSetpoint(-lift_limit);
+        }
+    } else {
+        this->arm_pid_->SetSetpoint(lift_limit);
+    }
 
     return true;
 }
 
 bool Arm::SetExtend(double inches) {
-    this->inches_pid_->SetSetpoint(inches);
+    OKC_CHECK(this->inches_pid_ != nullptr);
```

Comment:
this should be `arm_pid_` instead of `inches_pid_`

Reply:
Or something is inconsistent. But arm pid is being used in this function and being dereferenced so it probably should be that